bam_qc_metrics/bam_qc.py

```diff
@@ -748,20 +749,59 @@ def evaluate_bedtools_metrics(self):
             metrics['number of targets'] = targetBedTool.count()
             metrics['total target size'] = sum(len(f) for f in targetBedTool.features())
             metrics['reads on target'] = len(bamBedTool.intersect(self.target_path))
-            self.logger.debug("Found bedtools metrics")
+            result = self.evaluate_bedtools_coverage(bamBedTool, targetBedTool)
+            [metrics['total coverage'], metrics['coverage per target']] = result
+            self.logger.info("Found bedtools metrics")
         else:
             metrics['number of targets'] = None
             metrics['total target size'] = None
             metrics['reads on target'] = None
+            metrics['total coverage'] = None
+            metrics['coverage per target'] = {}
             self.logger.info("No target path given, omitting bedtools metrics")
-        # placeholders; TODO implement these metrics
-        metrics['total coverage'] = None
-        metrics['coverage per target'] = None
-        # TODO add bedtools coverage metrics?
-        #coverage = targetBedTool.coverage(self.bam_path)
-        #print(coverage)
         return metrics
 
+    def evaluate_bedtools_coverage(self, bamBedTool, targetBedTool):
+        """
+        Find the bedtools coverage histogram and process to get coverage by target.
+        For each target, coverage = total bases on target / target size
+        """
+        hist_str = self.run_bedtools()
+        self.logger.debug("Found bedtools coverage histogram")
+        reader = csv.reader(re.split("\n", hist_str.strip()), delimiter="\t")
+        all_name = 'all'
+        by_target = {}
+        target_sizes = {}
+        # row either starts with 'all', or with (chromosome, start, end)
+        # row ends with (depth, bases, target_size, fraction_covered)
+        int_expr = re.compile("^[0-9]+$")
+        for row in reader:
+            if len(row) == 5 and row[0] == all_name:
+                target = all_name
+            elif len(row) >= 7 and int_expr.match(row[1]) and int_expr.match(row[2]):
```

Comment:
If the target is named, it will be in the fourth column (the --hist result file will have >= 8 columns). If that's the case, `target` should be set to that name rather than the target position.

Reply:
OK, will fix.